conbench/entities/hardware.py

```diff
@@ -0,0 +1,203 @@
+
+    @classmethod
+    def create(cls, data):
+        hardware = cls(**data)
+        hardware.hash = hardware.generate_hash()
+        hardware.save()
+        return hardware
+
+
+class Machine(Hardware):
+    architecture_name = Nullable(s.Text)
+    kernel_name = Nullable(s.Text)
+    os_name = Nullable(s.Text)
+    os_version = Nullable(s.Text)
+    cpu_model_name = Nullable(s.Text)
+    cpu_l1d_cache_bytes = Nullable(s.Integer, check("cpu_l1d_cache_bytes>=0"))
+    cpu_l1i_cache_bytes = Nullable(s.Integer, check("cpu_l1i_cache_bytes>=0"))
+    cpu_l2_cache_bytes = Nullable(s.Integer, check("cpu_l2_cache_bytes>=0"))
+    cpu_l3_cache_bytes = Nullable(s.Integer, check("cpu_l3_cache_bytes>=0"))
+    cpu_core_count = Nullable(s.Integer, check("cpu_core_count>=0"))
+    cpu_thread_count = Nullable(s.Integer, check("cpu_thread_count>=0"))
+    cpu_frequency_max_hz = Nullable(s.BigInteger, check("cpu_frequency_max_hz>=0"))
+    memory_bytes = Nullable(s.BigInteger, check("memory_bytes>=0"))
+    gpu_count = Nullable(s.Integer, check("gpu_count>=0"), default=0)
+    gpu_product_names = Nullable(postgresql.ARRAY(s.Text), default=[])
+
+    __mapper_args__ = {"polymorphic_identity": "machine"}
+
+    @classmethod
+    def upsert(cls, **kwargs):
+        machine = cls.first(**kwargs)
+        if not machine:
+            machine = cls.create(kwargs)
+        return machine
+
+    def generate_hash(self):
+        return (
+            self.name
+            + "-"
+            + str(self.gpu_count)
+            + "-"
+            + str(self.cpu_core_count)
+            + "-"
+            + str(self.cpu_thread_count)
+            + "-"
+            + str(self.memory_bytes)
+        )
+
+    def serialize(self):
+        return {
+            "id": self.id,
+            "name": self.name,
+            "type": self.type,
+            "architecture_name": self.architecture_name,
+            "kernel_name": self.kernel_name,
+            "os_name": self.os_name,
+            "os_version": self.os_version,
+            "cpu_model_name": self.cpu_model_name,
+            "cpu_l1d_cache_bytes": self.cpu_l1d_cache_bytes,
+            "cpu_l1i_cache_bytes": self.cpu_l1i_cache_bytes,
+            "cpu_l2_cache_bytes": self.cpu_l2_cache_bytes,
+            "cpu_l3_cache_bytes": self.cpu_l3_cache_bytes,
+            "cpu_core_count": self.cpu_core_count,
+            "cpu_thread_count": self.cpu_thread_count,
+            "cpu_frequency_max_hz": self.cpu_frequency_max_hz,
+            "memory_bytes": self.memory_bytes,
+            "gpu_count": self.gpu_count,
+            "gpu_product_names": self.gpu_product_names,
+            "links": {
+                "list": f.url_for("api.hardwares", _external=True),
+                "self": f.url_for("api.hardware", hardware_id=self.id, _external=True),
+            },
+        }
+
+
+class Cluster(Hardware):
+    info = Nullable(postgresql.JSONB)
+    distribution_info = Nullable(postgresql.JSONB)
```

Comment:
This might be a bit bike shedding (so tell me to go away if it is!), but I want to get these names right since they'll be with us / require a migration to change otherwise...


What about `info` (for the hash one) and `optional_info`? We avoid `distribution` which might be confusion with things like linux distributions. And it makes it a bit clearer that the optional one is a little more ephemeral / won't disrupt history?

Reply:
Done.